### DISCUSSION TITLE
Is it safe to run upgrade.sh at any point?

### DISCUSSION THREAD

**Discussion by tomkivlin** [netbox-community/netbox#7946] [Q&A] [answered]
I am looking to build a Netbox AMI and was wondering if there is any harm in running `./upgrade.sh` every time an instance is created?

e.g. AMI is created with the binaries, and then the configuration settings are passed in user-data during the cloud-init stage of first boot.  If something needs to change, a new AMI is created and new instance booted.  We're using managed PostgreSQL, Redis and S3 storage from AWS for the stateful parts, so just the netbox app and web layers to host in the VMs.

I've seen the docker approach uses a slightly modified script with a check before running e.g. `manage.py migrate` but wondered if it mattered.
i.e. if I just run it, but there is nothing to migrate, it doesn't do anything.

Part of `docker-entrypoint.sh`:

https://github.com/netbox-community/netbox-docker/blob/9c66cc4c327d269875b8b00d92d4924ad4a0341e/docker/docker-entrypoint.sh#L39-L42

```sh
# Check if update is needed
if ! ./manage.py migrate --check >/dev/null 2>&1; then
  echo "⚙️ Applying database migrations"
  ./manage.py migrate --no-input
```

Similar part of `upgrade.sh`:

https://github.com/netbox-community/netbox/blob/86ada33577786140899c28792c8c26fc68d599c5/upgrade.sh#L64-L67

(For those that will no doubt ask - I would like to use the docker approach but don't have a k8s cluster to hand and don't want to learn ECS - immutable EC2 is a compromise but OK for now).

**Comment by candlerb** [comment, marked as answer]
If you run the full `./upgrade.sh` then it will destroy the existing virtualenv and recreate it.  There's no particular harm, but it will delay your instance startup while it unpacks a whole load of python packages and recreates the virtualenv.

Actually, there is a risk worth mentioning.  Because Netbox doesn't "pip freeze" its dependencies, rebuilding the virtualenv may download a newer and incompatible version of one of Netbox's indirect dependencies.  If you are running up-to-date Netbox then the risk is fairly low.  The older the Netbox that you're running, the higher the chance that you will get bitten by this.

**Comment by tomkivlin** [comment]
Many thanks @candlerb - that makes sense. 

I will share the repo here once it's done, but I suspect I'll use a modified version like the docker approach where the virtualenv is used if possible and only created if necessary.
